clusterctl/cmd/validate_cluster.go

```diff
@@ -17,26 +17,130 @@ limitations under the License.
 package cmd
 
 import (
+	"fmt"
+
 	"github.com/golang/glog"
 	"github.com/spf13/cobra"
-	"sigs.k8s.io/cluster-api/errors"
+	"k8s.io/api/core/v1"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/cluster-api/pkg/apis/cluster/common"
+	"sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+	"sigs.k8s.io/cluster-api/pkg/client/clientset_generated/clientset"
 )
 
+type ValidateClusterOptions struct {
+	ClusterName string
+	Namespace   string
+}
+
+var vco = &ValidateClusterOptions{}
+
 var validateClusterCmd = &cobra.Command{
 	Use:   "cluster",
 	Short: "Validate a cluster created by cluster API.",
 	Long:  `Validate a cluster created by cluster API.`,
 	Run: func(cmd *cobra.Command, args []string) {
-		if err := RunValidateCluster(); err != nil {
-			glog.Exit(err)
-		}
+		RunValidateCluster()
 	},
 }
 
 func init() {
+	validateClusterCmd.Flags().StringVarP(
+		&vco.ClusterName, "cluster-name", "", "",
+		"The name of the cluster to be validated. This can be omitted if there is only one cluster being managed in the given namespace.")
+	validateClusterCmd.Flags().StringVarP(
+		&vco.Namespace, "namespace", "n", v1.NamespaceDefault,
+		"If present, validate the cluster in the specified namespace.")
 	validateCmd.AddCommand(validateClusterCmd)
 }
 
-func RunValidateCluster() error {
-	return errors.NotImplementedError
+func RunValidateCluster() {
+	// TODO(wangzhen127): Also validate the cluster in addition to the cluster API objects.
+	cluster := validateClusterAPIObjects()
+
+	glog.Infof("Pass. The cluster %q in namespace %q is validated.", cluster.Name, vco.Namespace)
+}
+
+func validateClusterAPIObjects() *v1alpha1.Cluster {
```

Comment:
why does this function return a cluster? That doesn't seem like a result type I would expect from a validate function.

Reply:
Mostly because it needs to print cluster name at the end. And since it allows omitting cluster name, it needs to find the cluster first. I moved getting cluster out of the function now. So it currently only returns error.